NeoML/src/Dnn/Layers/DropoutLayer.cpp

```diff
@@ -99,32 +99,42 @@ void CDropoutLayer::RunOnce()
 void CDropoutLayer::BackwardOnce()
 {
 	// Backward pass is only possible when learning
-	NeoAssert( desc != 0 );
+	NeoAssert( desc != nullptr );
 
 	MathEngine().Dropout( *desc, outputDiffBlobs[0]->GetData(), inputDiffBlobs[0]->GetData() );
 
 	if( !GetDnn()->IsRecurrentMode() || GetDnn()->IsFirstSequencePos() ) {
 		// Clear the memory after the whole sequence is processed
-		destroyDropoutDesc();
+		disableDropoutDesc();
 	}
 }
 
 void CDropoutLayer::initDropoutDesc()
 {
-	if( desc == 0 ) {
-		desc = MathEngine().InitDropout( dropoutRate, isSpatial, isBatchwise, inputBlobs[0]->GetDesc(), outputBlobs[0]->GetDesc(),
-			GetDnn()->Random().Next() );
+	if(desc == nullptr) {
+		desc = MathEngine().InitDropout();
+	}
+
+	if (!desc->isValid) {
+		MathEngine().UpdateDropout(desc, dropoutRate, isSpatial, isBatchwise, inputBlobs[0]->GetDesc(), outputBlobs[0]->GetDesc(),
```

Comment:
If you move `InitDropout();` from constructor to method `Reshape()`, you can avoid update so many fields.
In the `Reshape()` you may fixate at least `dropoutRate, isSpatial, isBatchwise`.
Maybe you should use pointers to `CBlobDescs` to avoid constructing empty structures.